src/Act/Consistency.hs

```diff
@@ -116,3 +130,440 @@ checkCases (Act _ contracts) solver' smttimeout debug = do
 
       failMsg str model = render (red (pretty str) <> line <> model <> line) >> exitFailure
       errorMsg str = render (pretty str <> line) >> exitFailure
+
+
+--- ** Array Bounds Checking ** ---
+
+type ModelCtx = Reader Model
+
+mkBounds :: TypedExp -> Int -> [Exp ABoolean]
+mkBounds (TExp SInteger e) b = [LEQ nowhere (LitInt nowhere 0) e, LT nowhere e (LitInt nowhere $ toInteger b)]
+mkBounds _ _ = error "Internal Error: Expected Integral Index"
+
+mkRefBounds :: Ref a -> [Exp ABoolean]
+mkRefBounds (SArray _ ref _ tes) = concatMap (uncurry mkBounds) tes <> mkRefBounds ref
+mkRefBounds (SMapping _ ref _ _) = mkRefBounds ref
+mkRefBounds (SField _ ref _ _) = mkRefBounds ref
+mkRefBounds _ = []
+
+mkStorageBounds :: StorageLocation -> [Exp ABoolean]
+mkStorageBounds (SLoc _ (Item _ _ ref)) = mkRefBounds ref
+
+mkCalldataBounds :: CalldataLocation -> [Exp ABoolean]
+mkCalldataBounds (CLoc _ (Item _ _ ref)) = mkRefBounds ref
+
+-- TODO: There are locs that don't need to be checked, e.g. assignment locs cannot be out of bounds
+mkConstrArrayBoundsQuery :: Constructor -> (Id, [StorageLocation], [CalldataLocation], SMTExp, SolverInstance -> IO Model)
```

Comment:
Is it possible to factor out duplicate code in `mkConstrArrayBoundsQuery` and `mkBehvArrayBoundsQuery`?